### PR TITLE
[HOLD] Recreate acceptance and functional tests. Installs phpbuiltinserver.

### PR DESCRIPTION
**Pull request by khafa100**
### HOLD

This PR is on hold until we can sort out the installation / testing issues.

If not resolved by 2017-09-30, close this branch without merging.

---

Story: https://trello.com/c/AsxohMgO

I've recreated the functional and acceptance tests, and required phpbuiltinserver again. The functional tests run without issues, but the acceptance tests require the server to be running in order to work. I'm not 100% sure but I think it needs to connect to port 8222 before connecting to port 8777?

I wasn't able to identify why the acceptance tests weren't running correctly for either of you before unfortunately.

@tatwell Do you want me to try and automate running the server when running the test? Or are you fine with two steps when running tests? Like so:

```
php -S localhost:8222 -t ./public
codecept run
```

### CODE REVIEW

**Comment by tatwell** [comment]
@khafa100 Yes, this step needs to be removed:

    php -S localhost:8222 -t ./public

That's where we were before. Tests should be configured so that running this command will start local server for acceptance tests and run all tests successfully:

    codecept run

**Comment by khafa100** [comment]
@tatwell Let me know if you want me to make any other changes. I wasn't sure if I should add extra information somewhere such as that the default selenium port is 4444, or how to change the ports. I  didn't cause I thought it might be too much and we'd want to just stick with the default anyway. If the tests and server run properly for you then I'll go ahead and merge.

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (tatwell) [inline tests/acceptance/expectUserAndAdminLoginCest.php]: Formatting is a little funky. Include a space between methods and don't put opening bracket on new line:

```
# Bad
public function method_name() 
{
}

# Good
public function method_name() {
}
```

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (tatwell) [inline tests/acceptance/expectUserAndAdminLoginCest.php]: Is this test testing anything? That is, would this line fail if credentials were wrong?

**Review by tatwell — COMMENTED** [review]
Looks good but I'll want to run tests before I approve. Will update when I've had a chance.

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (khafa100) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: I just tested it twice. Once by changing it to `$I->loginAs('chappresss', 'fail');` and `$I->loginAs('random', 'password');`

It passed both times. I'll look into why it doesn't fail when credentials are wrong and fix it so that it only passes with the correct credentials.

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (khafa100) [inline tests/acceptance/expectUserAndAdminLoginCest.php]: Hey @spence-man, is there a reason why this is just `function` and not `public function` (or private)? It works properly and I saw an example of this [here](http://codeception.com/docs/07-AdvancedUsage#Dependency-Injection) in the last block of code of the Dependency Injection section. I'm just wondering for my own knowledge. Is it because it defaults to public (or private)?

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (tatwell) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: Good, yes that's what I was wondering.

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (spence-man) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: Oh, I think that was unintentional on my part. It probably should have the `public` included just like the other ones.

Comment (spence-man) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: I'm not sure about the default behavior, sorry.

Comment (tatwell) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: Yes, php defaults to public for methods. I don't really care which way we do it so much as we're consistent about it.

But for questions like this, let's refer to this (until we get our own style guide page):

- http://www.phptherightway.com/

**Comment by tatwell** [comment]
@khafa100 On extra information question above, if it's not necessary for a developer to know to work with this, then don't put it in README. If not sure, put it in a comment on the story card. That way, if it's really needed some day, we can probably track it down.

If you think it will need to be referenced at times, but infrequently, you might add it to a repo wiki page.

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (khafa100) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: Ok, thanks both of you for clarifying and for the link!

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (khafa100) [inline tests/acceptance/expectUserAndAdminLoginCest.php]: I think there's a better way to do this, but when I used `$I->amOnPage(/wp-admin/profile.php);` it still passed regardless.

File: tests/acceptance/expectUserAndAdminLoginCest.php
Comment (khafa100) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: I tested this with a non-admin user I created locally, but I wasn't sure how to update `chappress_dev_dump`the correct way so that it would generate the database for other devs with the non-admin user I created.

Comment (tatwell) [reply tests/acceptance/expectUserAndAdminLoginCest.php]: It should use fixtures, like Rails, which create the user at the start of the test and then clear the database at the end.

**Comment by tatwell** [comment]
When I clone repo and follow instructions from scratch, tests are failing:

```
ERRORS!
Tests: 9, Assertions: 5, Errors: 4, Failures: 1
```

Here's the error:

> [ConnectionException] Can't connect to Webdriver at http://127.0.0.1:4444/wd/hub. Please make sure that Selenium Server or PhantomJS is running.

Note: I am not starting the dev server when I go through README install steps. I do install and start the Selenium Server as documented.

When you are done with updates, make sure you test documented steps with fresh clone.

**Review by tatwell — COMMENTED** [review]
@khafa100 I've reorganized the README to better reflect how I expect things to work when a dev starts from scratch.

This is going to create a conflict with @spence-man's changes once these branches get merged together. Whoever ends up being responsible for resolving the conflict, please make sure that these changes are preserved.

File: README.md
Comment (tatwell) [inline README.md]: This keeps coming back but I really don't want this. We need to automate this. New dev should be able to go directly from command-line install steps to running tests.

**Comment by khafa100** [comment]
@tatwell For the error you mentioned:

> [ConnectionException] Can't connect to Webdriver at http://127.0.0.1:4444/wd/hub. Please make sure that Selenium Server or PhantomJS is running.

I was able to recreate it and solve it with `brew services restart selenium-server-standalone`. I documented it in the `README`. If it still doesn't work for you though, let me know.

File: README.md
Comment (khafa100) [reply README.md]: I've automated it with:
```
wp core install --url=http://localhost:8222/ --title=chap-press --admin_user=chappress --admin_password=password --admin_email=chappress@gmail.com
```
The issue I brought up to you earlier was because I accidentally ran this command in the root folder instead of the `public` folder. So it does exactly as you asked and the config works exactly as it should.

File: README.md
Comment (tatwell) [inline README.md]: This should be under a new bullet point.

Comment (tatwell) [reply README.md]: Steps missing:

```
$ wp core install --url=http://localhost:8222/ --title=chap-press --admin_user=chappress --admin_password=password --admin_email=chappress@gmail.com
-bash: wp: command not found
```

**Review by tatwell — CHANGES_REQUESTED** [review]

File: README.md
Comment (tatwell) [inline README.md]: Formatting needs to be fixed. See here:

- https://github.com/chapmanu/chap-press/tree/recreate-tests

Comment (tatwell) [reply README.md]: Also title is off. Maybe call it:

- Initialize Database for Local Server

Comment (tatwell) [reply README.md]: Not sure if this works or not. This is what I got after installing `wp-cli`:

```
$ wp core install --url=http://localhost:8222/ --title=chap-press --admin_user=chappress --admin_password=password --admin_email=chappress@gmail.com
Error: This does not seem to be a WordPress install.
Pass --path=`path/to/wordpress` or run `wp core download`.
SMC-MBOOKP13-01:chap-press atwell$ cd public/
SMC-MBOOKP13-01:public atwell$ wp core install --url=http://localhost:8222/ --title=chap-press --admin_user=chappress --admin_password=password --admin_email=chappress@gmail.com
WordPress is already installed.
```

I'd prefer everything was run from project root without having to change directories.

Comment (tatwell) [reply README.md]: Hold on. This may be my fault. I forgot to switch branches. Let me start over.

Comment (tatwell) [reply README.md]: No. Same result when I change branches. And I still get the original error:

>  [ConnectionException] Can't connect to Webdriver at http://127.0.0.1:4444/wd/hub. Please make sure that Selenium Server or PhantomJS is running.  

I am going to put this PR on hold for now.

**Review by spence-man — APPROVED** [review]
Issues resolved at [PR #21](https://github.com/chapmanu/chap-press/pull/21).

**Comment by spence-man** [comment]
Issue resolved on PR #21 . Close this branch without merging

**Comment by spence-man** [comment]
@tatwell Can I go ahead and close this and not merge? [PR #21](https://github.com/chapmanu/chap-press/pull/21) took care of this.

**Comment by tatwell** [comment]
@spence-man Yes, please close with short comment explaining why. Thanks.

**Comment by spence-man** [comment]
This PR issue was settled by a more updated PR #21 . Many changes were made since this PR was opened so a new PR was opened at a later date to avoid merge conflicts with the development branch. The acceptance and functional tests are able to functional on a separate port using Codeception extension `Run Process` & using `WPWebDriver`. See the PR  for more information

**Pull request closed without merging**
